backend/backend.go

```diff
@@ -122,3 +123,63 @@ func UnmarshalHistoryEvent(bytes []byte) (*HistoryEvent, error) {
 	}
 	return e, nil
 }
+
+func purgeOrchestrationState(ctx context.Context, be Backend, iid api.InstanceID, recursive bool) (int, error) {
+	owi := &OrchestrationWorkItem{
+		InstanceID: iid,
+	}
+	state, err := be.GetOrchestrationRuntimeState(ctx, owi)
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch orchestration state: %w", err)
+	}
+	if len(state.NewEvents())+len(state.oldEvents) == 0 {
+		// If there are no events, the orchestration instance doesn't exist
+		return 0, api.ErrInstanceNotFound
+	}
+	if !state.IsCompleted() {
+		// Orchestration must be completed before purging its state
+		return 0, api.ErrNotCompleted
+	}
+	deletedInstanceCount := 0
+	subOrchestrationInstances := getSubOrchestrationInstances(state.OldEvents(), state.NewEvents())
+	for _, subOrchestrationInstance := range subOrchestrationInstances {
+		count, err := purgeOrchestrationState(ctx, be, subOrchestrationInstance, recursive)
+		deletedInstanceCount += count
```

Comment:
Shouldn't we increment `deletedInstanceCount` only if it succeeds?

Reply:
If `purgeOrchestrationState` returns `count` and `err` that means `count` orchestrations(sub-orchestrations) have been successfully purged, and it errored out on purging the next one. e.g. If an orchestration has 10 sub-orchestrations, and 5 got purged, but failed to purge 6th one, We are returning count as 5 and error as received in purging the 6th one.